{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "091fea42-9b1d-4535-802f-6015d5891f48",
   "metadata": {},
   "source": [
    "# Miscellaneous code and commands used in our study of intrinsic resistance in enterococci\n",
    "\n",
    "### Data retrieval, preprocessing and quality checks\n",
    "\n",
    "- Dowload metadata from *Enterococcus* genome entries in the NCBI Assembly database\n",
    "\n",
    "```bash\n",
    "./utils/download_ncbi_assembly_metadata.sh enterococcus enterococcus_ncbi_assembly_metadata.tsv\n",
    "```\n",
    "\n",
    "- Download the assembly fasta and gff files\n",
    "\n",
    "```python\n",
    "import os\n",
    "with open('data/enterococcus_ncbi_assembly_metadata.tsv', 'r') as f:\n",
    "    for line in f:\n",
    "        flds = line.strip().split('\\t')\n",
    "        acc = flds[6]\n",
    "        asm = flds[4]\n",
    "        asm_name = acc + '_' + asm\n",
    "        ftp_url = flds[17]\n",
    "        fna_url = ftp_url + '/' + asm_name + '_' + '_genomic.fna.gz'\n",
    "        faa_url = ftp_url + '/' + asm_name + '_' + '_protein.faa.gz'\n",
    "        gff_url = ftp_url + '/' + asm_name + '_' + '_genomic.gff.gz'\n",
    "\n",
    "        os.system('wget {} -P data/ncbi/genome'.format(fna_url))\n",
    "        os.system('wget {} -P data/ncbi/protein'.format(faa_url))\n",
    "        os.system('wget {} -P data/ncbi/annotation'.format(gff_url))\n",
    "```\n",
    "\n",
    "- Run CheckM to calculate assembly metrics\n",
    "```bash\n",
    "checkm analyze firmicutes.ms data/ncbi/genome checkm_out -t 8 --ali --nt -x fa\n",
    "```\n",
    "\n",
    "- Run HMMer against the Phylosift database\n",
    "```python\n",
    "import os\n",
    "for f in os.listdir('data/ncbi/protein'):\n",
    "    file_name = os.path.basename(f)\n",
    "    genome = file_name.replace('.faa', '')\n",
    "    cmd = \"hmmsearch --tblout output/phylosift_out/{}.out data/PhyloSift/PhyloSift.hmm {}\".format(genome, f)\n",
    "    os.system(cmd)\n",
    "```\n",
    "\n",
    "### Genome annotation \n",
    "\n",
    "- Run prokka on genome assemblies with *Enterococcus* specific reference genes```\n",
    "\n",
    "```python\n",
    "import os\n",
    "for f in os.listdir('data/ncbi/genome'):\n",
    "    fna_file = os.path.basename(f)\n",
    "    genome = fna_file.replace('.fna', '')\n",
    "    cmd = 'prokka --outdir {} --prefix {} --centre X --addgenes --locustag {} --cpus 12 --compliant --proteins data/enterococcus_refgenes.faa {}'.format(genome, genome, genome, fna_file)\n",
    "    os.system(cmd)\n",
    "```\n",
    "\n",
    "- Run HMMer on the PFAM database (release 32)\n",
    "\n",
    "```bash\n",
    "hmmscan -E 1 --cpu 2 --domtblout genome.pfam.out data/pfam/Pfam-A.hmm data/annotation/prokka/genome/genome.faa \n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "13477428-f97f-4dc8-8801-6388c8615ffb",
   "metadata": {},
   "source": [
    "### Build SCC phylogeny\n",
    "\n",
    "- Run OrthoFinder\n",
    "```bash\n",
    "orthofinder -f output/orthofinder/input -t 32 -og\n",
    "```\n",
    "- Create SCC gene alignments and concatenate them\n",
    "```python\n",
    "./build_scc_phylogeny.py\n",
    "```\n",
    "\n",
    "- Run IQTREE on the SCC genes alignment file, use MFP to find the best model and calculate bootstrap values\n",
    "```bash\n",
    "iqtree -s SCCorthogroups_aligned_nuc.fasta --prefix iqtree_mfp -bb 1000 -wbtl -nt 32 -m MFP --seed 1 -spp SCCorthogroups_partition_file.txt\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "59f30e7d-cd0d-4a72-8bc0-ba8915b42c26",
   "metadata": {},
   "source": [
    "## Find mobile genetic elements \n",
    "\n",
    "To make sure we find as many mobile elements as we can, we made use of several methods and databases.\n",
    "\n",
    "- TnFinder Database\n",
    "\n",
    "We used the `tncentraldb` and `Tn3R` databases, and ran `blastn`\n",
    "```bash\n",
    "blastn -db data/tncentraldb/tncentraldb -query data/genome/genome.fna \\\n",
    "    -outfmt '6 qaccver ssciname stitle saccver pident length mismatch gapopen qstart qend sstart send evalue bitscore qcovs' \\\n",
    "    -evalue 0.001 -num_threads 4 -out output/mge/blastn_tncentraldb_genome.out\n",
    "```\n",
    "- ISFinder Database\n",
    "\n",
    "We used the ISfinder database and ran `blastn`\n",
    "``` bash\n",
    "blastn -db data/isfinderdb/IS -query data/genome/genome.fna \\\n",
    "    -outfmt '6 qaccver ssciname stitle saccver pident length mismatch gapopen qstart qend sstart send evalue bitscore qcovs' \\\n",
    "    -evalue 0.001 -num_threads 4 -out output/mge/blastn_isfinderis_genome.out\n",
    "```\n",
    "\n",
    "- Ran MOBsuite (v3.0.3)\n",
    "\n",
    "Since our assembly sequences are made up of multiple contigs/scaffolds, we used the `--multi` option of `mob_typer`\n",
    "```bash\n",
    "mob_typer --multi --infile data/genome/genoma.fna --out_file output/mge/mob_genome.out -n 2\n",
    "```\n",
    "\n",
    "- Ran PlasmidFinder (v2.1.1)\n",
    "\n",
    "``` bash\n",
    "plasmidfinder.py -i data/genome/genome.fna -o output/mge/plasmidfinder -x -p data/plasmidfinderdb\n",
    "```\n",
    "- DeepMicrobe\n",
    "\n",
    "```bash\n",
    "python bin/DeepMicrobeFinder/predict.py -i data/genome/genome.fna one-hot -d data/DeepMicrobeFinder/models/one-hot-models/ -m single -o genome -l 2000\n",
    "```\n",
    "\n",
    "- Ran Prophet (v0.5.1) to predict phage sequences\n",
    "\n",
    "```bash\n",
    "perl bin/ProphET-0.5.1/ProphET_standalone.pl --fasta_in data/genome/genome.fna --gff_in data/annotaion/prokka/genome.gff --outdir output/mge/prophet\n",
    "```"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (Spyder)",
   "language": "python3",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
